### DISCUSSION TITLE
[Proposal]: A checked arithmetic expression on ushort/ubyte/byte can be lifted or inferred to return ushort/ubyte/byte in a checked context.

### DISCUSSION THREAD

**Discussion by otac0n** [dotnet/csharplang#7710] [Language Ideas]
* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary

Checked expressions can use type inference to avoid compile-time errors.

## Motivation

Subtracting from unsigned or small integer types currently requires a cast.  Checked expressions already provide guarantees that the type system needs.

<img width="119" alt="image" src="https://github.com/dotnet/csharplang/assets/193584/7e2d47ff-c806-44ca-9614-9859e0df5762">

```
ushort x = 3;
x = checked(x - 1);
```

Currently this fails with a `CS0266` error.  Allowing such expressions would thus be enabling a previously illegal statement, which is non-breaking from a language-only perspective.

## Design Meetings

None yet

**Comment by otac0n** [comment]
This is similar and particularly frustrating:
<img width="159" alt="image" src="https://github.com/dotnet/csharplang/assets/193584/068fd32c-9a5d-4cf8-a82f-97642f3b3ce7">
```
ushort x = 3;
x = x switch
{
    > 0 => checked(x - 1),
    _ => x,
};
```

**Comment by colejohnson66** [comment]
IIRC, this has to do with how the CIL encodes arithmetic operations (everything is upcast to an int or larger), not overflow support.

**Reply by otac0n** [reply]
You are basically correct, buy I'm proposing a change in semantics that can add the casts for you.
Lowering `x = checked(x - 1);` to `x = checked((ushort)(x - 1));` in the case of what would today be an error.
It would require a change in semantics for arithmetic operations, but can be made backwards compatible with existing code by only applying in today's error case.

**Comment by huoyaoyuan** [comment]
`checked(ushort.MaxValue + 1)` is valid and will return 65536, without throwing. `checked(short.MinValue - 1)` will also return an int32.

**Reply by otac0n** [reply]
Yes, it would be `checked((ushort)(ushort.MaxValue + 1))`. Although you will probably hit a compile-time constant folding error rather than an AritmeticOverflow.

**Comment by otac0n** [comment]
If, in a checked context, you add a `checked()` wrapper around an arithmetic expression that would be forced to `int`, can we instead lift the expression to the lowest inferred type?
